### PR TITLE
Add hotkeys for save areas of the map

### LINKED ISSUE #2956: Save areas of the map
Here is a mockup image I have made :
744383ScreenShot001

(note that to fit the minimap size, buttons (which look like tabs) should be a bit smaller)

There could be new icons next to the minimap:

- clicking on the " + " button would record the current position of the view and add a new "button" with a digit to quickly go to this position later

- clicking on the " - " button would remove a previously recorded position (the player would have to select it first before being able to click on the " - " button)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/136843-save-areas-of-the-map?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by viszkoktamas**
Closes #2956 Add hotkeys for save areas of the map, and hotkeys for jump to saved positions

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 1]
```diff
@@ -72,6 +72,11 @@ public class ViewportControllerWidget : Widget
 		ScrollDirection edgeDirections;
 		World world;
 		WorldRenderer worldRenderer;
+		WPos viewPortBookmarkSlot1;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: You could cut out a lot of duplication by making this an array of WPos and then extracting helper methods for `SaveBookmark(int index, WPos position)` and `JumpToBookmark(int index)`

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 2]
```diff
@@ -305,6 +310,66 @@ public override bool HandleKeyPress(KeyInput e)
...
+			if (key == ks.ViewPortBookmarkSaveSlot4)
+			{
+				viewPortBookmarkSlot4 = worldRenderer.Viewport.CenterPosition;
+				return false;
+			}
+
+			if (key == ks.ViewPortBookmarkUseSlot1)
+			{
+				if (viewPortBookmarkSlot1.X != 0)
+				{
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: Please note that our [code style](https://github.com/OpenRA/OpenRA/wiki/Coding-Standard) is to omit braces around single line blocks.

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 3]
```diff
@@ -305,6 +310,66 @@ public override bool HandleKeyPress(KeyInput e)
...
+
+			if (key == ks.ViewPortBookmarkSaveSlot4)
+			{
+				viewPortBookmarkSlot4 = worldRenderer.Viewport.CenterPosition;
+				return false;
+			}
+
+			if (key == ks.ViewPortBookmarkUseSlot1)
+			{
+				if (viewPortBookmarkSlot1.X != 0)
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: Using a magic value of `x == 0` isn't a great idea, since 0 is potentially a valid value.  A more robust approach is to use `WPos?` and then compare against null.

File: OpenRA.Game/Settings.cs [hunk 4]
```diff
@@ -207,6 +207,17 @@ public class KeySettings
...
+		public Hotkey ViewPortBookmarkSaveSlot2 = new Hotkey(Keycode.NUMBER_2, Modifiers.Alt | Modifiers.Shift);
+		public Hotkey ViewPortBookmarkSaveSlot3 = new Hotkey(Keycode.NUMBER_3, Modifiers.Alt | Modifiers.Shift);
+		public Hotkey ViewPortBookmarkSaveSlot4 = new Hotkey(Keycode.NUMBER_4, Modifiers.Alt | Modifiers.Shift);
+
+		public Hotkey ViewPortBookmarkUseSlot1 = new Hotkey(Keycode.NUMBER_1, Modifiers.Alt);
+		public Hotkey ViewPortBookmarkUseSlot2 = new Hotkey(Keycode.NUMBER_2, Modifiers.Alt);
+		public Hotkey ViewPortBookmarkUseSlot3 = new Hotkey(Keycode.NUMBER_3, Modifiers.Alt);
+		public Hotkey ViewPortBookmarkUseSlot4 = new Hotkey(Keycode.NUMBER_4, Modifiers.Alt);
+
+
```
Comment (pchote) [inline OpenRA.Game/Settings.cs]: Multiple blank lines is another code style issue.

File: OpenRA.Game/Settings.cs [hunk 5]
```diff
@@ -207,6 +207,17 @@ public class KeySettings
 		public Hotkey MapScrollLeft = new Hotkey(Keycode.LEFT, Modifiers.None);
 		public Hotkey MapScrollRight = new Hotkey(Keycode.RIGHT, Modifiers.None);
 
+		public Hotkey ViewPortBookmarkSaveSlot1 = new Hotkey(Keycode.NUMBER_1, Modifiers.Alt | Modifiers.Shift);
+		public Hotkey ViewPortBookmarkSaveSlot2 = new Hotkey(Keycode.NUMBER_2, Modifiers.Alt | Modifiers.Shift);
+		public Hotkey ViewPortBookmarkSaveSlot3 = new Hotkey(Keycode.NUMBER_3, Modifiers.Alt | Modifiers.Shift);
+		public Hotkey ViewPortBookmarkSaveSlot4 = new Hotkey(Keycode.NUMBER_4, Modifiers.Alt | Modifiers.Shift);
+
+		public Hotkey ViewPortBookmarkUseSlot1 = new Hotkey(Keycode.NUMBER_1, Modifiers.Alt);
```
Comment (pchote) [inline OpenRA.Game/Settings.cs]: This will conflict with the control groups: alt + number is used to center the viewport on the numbered group.

**Comment by pchote** [comment]
This is a feature i've wanted to see for a while, so thanks for taking it on!

I've left a few inline code comments that will need to be addressed before we take this.

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs (hunk 3 above)
Comment (LipkeGu) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: @viszkoktamas keep in mind it seams you mean ```null``` (HEX: 0x00)  instead of ```0``` (HEX: 0x30)
the different is the first "0" is nothing and the second one is a printable char which is used to terminate strings (for example) :)

**Comment by LipkeGu** [comment]
@viszkoktamas You can run ```make check``` to check for code style errors

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs (hunk 3 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: @LipkeGu How is that relevant here?

**Comment by viszkoktamas** [comment]
@LipkeGu Thank you for your advice, i did run check and fixed code style errors.

**Comment by viszkoktamas** [comment]
When i sqash my commits should i click on Update branch or not? I ask this because i did not commit "	Merge branch 'bleed' into viewportBookmark".

**Comment by viszkoktamas** [comment]
@pchote Im not sure about whats wrong with my new code. Maybe i clicked on "Update branch" button when cheks were running.

**Comment by reaperrr** [comment]
@viszkoktamas I like the feature and it works fine for Q, W and E, but it always crashes on Ctrl + R for me:
```Red Alert Mod at Version {DEV_VERSION}
on map 602f16e64f2e10d4a11f5709348d6a1e923f2719 (A Path Beyond by Westwood Studios).
Operating System: Windows (Microsoft Windows NT 6.1.7601 Service Pack 1)
Runtime Version: .NET CLR 4.0.30319.34209
Exception of type `System.IndexOutOfRangeException`: Der Index war außerhalb des Arraybereichs.
   bei OpenRA.Mods.Common.Widgets.ViewportControllerWidget.SaveBookmark(Int32 index, WPos position)
   bei OpenRA.Mods.Common.Widgets.ViewportControllerWidget.SaveCurrentPositionToBookmark(Int32 index)
   bei OpenRA.Mods.Common.Widgets.ViewportControllerWidget.HandleKeyPress(KeyInput e)
   bei OpenRA.Widgets.Widget.HandleKeyPressOuter(KeyInput e) in e:\ORA_work_reviewing\OpenRA.Game\Widgets\Widget.cs:Zeile 393.
   bei OpenRA.Widgets.Widget.HandleKeyPressOuter(KeyInput e) in e:\ORA_work_reviewing\OpenRA.Game\Widgets\Widget.cs:Zeile 389.
   bei OpenRA.Widgets.Widget.HandleKeyPressOuter(KeyInput e) in e:\ORA_work_reviewing\OpenRA.Game\Widgets\Widget.cs:Zeile 389.
   bei OpenRA.Widgets.Widget.HandleKeyPressOuter(KeyInput e) in e:\ORA_work_reviewing\OpenRA.Game\Widgets\Widget.cs:Zeile 389.
   bei OpenRA.Widgets.Ui.HandleKeyPress(KeyInput e) in e:\ORA_work_reviewing\OpenRA.Game\Widgets\Widget.cs:Zeile 120.
   bei OpenRA.DefaultInputHandler.<>c__DisplayClass1.<OnKeyInput>b__0() in e:\ORA_work_reviewing\OpenRA.Game\Input\InputHandler.cs:Zeile 40.
   bei OpenRA.Sync.CheckSyncUnchanged[T](World world, Func`1 fn) in e:\ORA_work_reviewing\OpenRA.Game\Sync.cs:Zeile 183.
   bei OpenRA.DefaultInputHandler.OnKeyInput(KeyInput input) in e:\ORA_work_reviewing\OpenRA.Game\Input\InputHandler.cs:Zeile 40.
   bei OpenRA.Platforms.Default.Sdl2Input.PumpInput(IInputHandler inputHandler) in e:\ORA_work_reviewing\OpenRA.Platforms.Default\Sdl2Input.cs:Zeile 166.
   bei OpenRA.Platforms.Default.Sdl2GraphicsDevice.PumpInput(IInputHandler inputHandler) in e:\ORA_work_reviewing\OpenRA.Platforms.Default\Sdl2GraphicsDevice.cs:Zeile 354.
   bei OpenRA.Renderer.EndFrame(IInputHandler inputHandler) in e:\ORA_work_reviewing\OpenRA.Game\Renderer.cs:Zeile 170.
   bei OpenRA.Game.RenderTick() in e:\ORA_work_reviewing\OpenRA.Game\Game.cs:Zeile 586.
   bei OpenRA.Game.Loop() in e:\ORA_work_reviewing\OpenRA.Game\Game.cs:Zeile 696.
   bei OpenRA.Game.Run() in e:\ORA_work_reviewing\OpenRA.Game\Game.cs:Zeile 714.
   bei OpenRA.Program.Run(String[] args) in e:\ORA_work_reviewing\OpenRA.Game\Support\Program.cs:Zeile 118.
   bei OpenRA.Program.Main(String[] args) in e:\ORA_work_reviewing\OpenRA.Game\Support\Program.cs:Zeile 40.```

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 6]
```diff
@@ -305,6 +328,54 @@ public override bool HandleKeyPress(KeyInput e)
 				return true;
 			}
 
+			if (key == ks.ViewPortBookmarkSaveSlot1)
+			{
+				SaveCurrentPositionToBookmark(1);
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: 0

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 7]
```diff
@@ -305,6 +328,54 @@ public override bool HandleKeyPress(KeyInput e)
...
+				return false;
+			}
+
+			if (key == ks.ViewPortBookmarkSaveSlot2)
+			{
+				SaveCurrentPositionToBookmark(2);
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: 1

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 8]
```diff
@@ -305,6 +328,54 @@ public override bool HandleKeyPress(KeyInput e)
...
+				return false;
+			}
+
+			if (key == ks.ViewPortBookmarkSaveSlot3)
+			{
+				SaveCurrentPositionToBookmark(3);
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: 2

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 9]
```diff
@@ -305,6 +328,54 @@ public override bool HandleKeyPress(KeyInput e)
...
+				return false;
+			}
+
+			if (key == ks.ViewPortBookmarkSaveSlot4)
+			{
+				SaveCurrentPositionToBookmark(4);
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: 3

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 10]
```diff
@@ -305,6 +328,54 @@ public override bool HandleKeyPress(KeyInput e)
...
+				return false;
+			}
+
+			if (key == ks.ViewPortBookmarkUseSlot1)
+			{
+				JumpToSavedBookmark(1);
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: And again.

**Comment by viszkoktamas** [comment]
@reaperrr thanks for feedback, i fixed the problem.

**Comment by GraionDilach** [comment]
:+1:

**Comment by reaperrr** [comment]
:+1:

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/035dabc10cce3ed23d9f3f8221f0910ed3926b32...732d437dce79e08ce67a295c264f068993c35178)
